extensions/druid-rocketmq/src/main/java/io/druid/firehose/rocketmq/RocketMQFirehoseFactory.java

```diff
@@ -0,0 +1,571 @@
+      synchronized (this) {
+        this.requestsWrite.add(request);
+        if (!hasNotified) {
+          hasNotified = true;
+          notify();
+        }
+      }
+    }
+
+    private void swapRequests()
+    {
+      List<DruidPullRequest> tmp = requestsWrite;
+      requestsWrite = requestsRead;
+      requestsRead = tmp;
+    }
+
+    @Override
+    public String getServiceName()
+    {
+      return getClass().getSimpleName();
+    }
+
+    /**
+     * Core message pulling logic code goes here.
+     */
+    private void doPull()
+    {
+      for (DruidPullRequest pullRequest : requestsRead) {
+        PullResult pullResult;
+        try {
+          if (!pullRequest.isLongPull()) {
+            pullResult = defaultMQPullConsumer.pull(
+                pullRequest.getMessageQueue(),
+                pullRequest.getTag(),
+                pullRequest.getNextBeginOffset(),
+                pullRequest.getPullBatchSize()
+            );
+          } else {
+            pullResult = defaultMQPullConsumer.pullBlockIfNotFound(
+                pullRequest.getMessageQueue(),
+                pullRequest.getTag(),
+                pullRequest.getNextBeginOffset(),
+                pullRequest.getPullBatchSize()
+            );
+          }
+
+          switch (pullResult.getPullStatus()) {
+            case FOUND:
+              // Handle pull result.
+              if (!messageQueueTreeSetMap.keySet().contains(pullRequest.getMessageQueue())) {
+                messageQueueTreeSetMap.putIfAbsent(
+                    pullRequest.getMessageQueue(),
+                    new ConcurrentSkipListSet<>(new MessageComparator())
+                );
+              }
+              messageQueueTreeSetMap.get(pullRequest.getMessageQueue()).addAll(pullResult.getMsgFoundList());
+              break;
+
+            case NO_NEW_MSG:
+            case NO_MATCHED_MSG:
+              break;
+
+            case OFFSET_ILLEGAL:
+              LOGGER.error(
+                  "Bad Pull Request: Offset is illegal. Offset used: {}",
+                  pullRequest.getNextBeginOffset()
+              );
+              break;
+
+            default:
+              break;
+          }
+        }
+        catch (MQClientException | RemotingException | MQBrokerException | InterruptedException e) {
+          LOGGER.error("Failed to pull message from broker.", e);
+        }
+        finally {
+          pullRequest.getCountDownLatch().countDown();
+        }
+
+      }
+      requestsRead.clear();
+    }
+
+    /**
+     * Thread looping entry.
+     */
+    @Override
+    public void run()
+    {
+      LOGGER.info(getServiceName() + " starts.");
+      while (!isStoped()) {
+        waitForRunning(0);
+        doPull();
+      }
+
+      // in case this service is shutdown gracefully without interruption.
+      try {
+        Thread.sleep(10);
+      }
+      catch (InterruptedException e) {
+        LOGGER.error("", e);
+      }
+
+      synchronized (this) {
```

Comment:
Move synchronization to Object within the DruidPullMessageService to prevent someone doing something screwy like synchronizing on a DruidPullMessageService instance.